File: tarea_1/1.4_matrices/cpp/src/main.cpp
#include "../include/RedistribucionController.h"
#include "../include/RedistribucionService.h"
#include "../include/OptimizadorGreedy.h"
#include <iostream>

using namespace std;

class RedistribucionUI {
private:
    RedistribucionController controller;

public:
    RedistribucionUI() : controller(make_unique<RedistribucionService>(make_unique<OptimizadorGreedy>())) {}

    void ejecutar() {
        int opcion;
        do {
            mostrarMenu();
            cin >> opcion;
            try {
                procesarOpcion(opcion);
            } catch (const exception& e) {
                cout << "❌ Error: " << e.what() << endl;
            }
        } while (opcion != 0);
    }

private:
    void mostrarMenu() {
        cout << "\n=== REASIGNACIÓN ÓPTIMA DE PRODUCTOS ===\n"
                  << "1. Configurar matrices manualmente\n"
                  << "2. Cargar datos de prueba\n"
                  << "3. Ejecutar optimización\n"
                  << "4. Mostrar resultados\n"
                  << "0. Salir\nOpción: ";
    }

    void procesarOpcion(int opcion) {
        switch (opcion) {
            case 1: controller.configurarMatrices(); break;
            case 2: controller.cargarDatosPrueba(); break;
            case 3: controller.ejecutarOptimizacion(); break;
            case 4: controller.mostrarResultados(); break;
            case 0: cout << "👋 Saliendo del sistema...\n"; break;
            default: cout << "❌ Opción inválida\n";
        }
    }
};

int main() {
    cout << "🚚 Sistema de Reasignación Óptima de Productos\n";
    
    RedistribucionUI app;
    app.ejecutar();
    
    return 0;
}

File: tarea_1/1.4_matrices/python/main.py
#!/usr/bin/env python3

from optimizador_greedy import OptimizadorGreedy
from redistribucion_service import RedistribucionService

class RedistribucionUI:
    def __init__(self):
        optimizador = OptimizadorGreedy()
        self.service = RedistribucionService(optimizador)
    
    def ejecutar(self):
        while True:
            self.mostrar_menu()
            try:
                opcion = int(input("Opción: "))
                if opcion == 0:
                    print("👋 Saliendo del sistema...")
                    break
                self.procesar_opcion(opcion)
            except ValueError:
                print("❌ Entrada inválida. Ingrese un número.")
            except Exception as e:
                print(f"❌ Error: {e}")
    
    def mostrar_menu(self):
        print("\n=== REASIGNACIÓN ÓPTIMA DE PRODUCTOS ===")
        print("1. Configurar matrices manualmente")
        print("2. Cargar datos de prueba")
        print("3. Ejecutar optimización")
        print("4. Mostrar resultados")
        print("0. Salir")
    
    def procesar_opcion(self, opcion: int):
        if opcion == 1:
            self.service.configurar_matrices()
        elif opcion == 2:
            self.service.cargar_datos_prueba()
        elif opcion == 3:
            self.service.ejecutar_optimizacion()
        elif opcion == 4:
            self.service.mostrar_resultados()
        else:
            print("❌ Opción inválida")

def main():
    print("🚚 Sistema de Reasignación Óptima de Productos (Python)")
    
    app = RedistribucionUI()
    app.ejecutar()

if __name__ == "__main__":
    main()